### PR TITLE
Kernel: Ensure network adapters use virtual memory correctly

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

File: Kernel/Net/RTL8139NetworkAdapter.cpp [hunk 1]
```diff
@@ -339,31 +338,31 @@ void RTL8139NetworkAdapter::send_raw(const u8* data, size_t length)
...
+        klog() << "RTL8139NetworkAdapter::receive got bad packet status=0x" << String::format("%x", status) << " length=" << length;
         reset();
         return;
     }
 
     // we never have to worry about the packet wrapping around the buffer,
     // since we set RXCFG_WRAP_INHIBIT, which allows the rtl8139 to write data
     // past the end of the alloted space.
-    memcpy((u8*)m_packet_buffer, (const u8*)(start_of_packet + 4), length - 4);
+    memcpy((u8*)m_packet_buffer->vaddr().as_ptr(), (const u8*)(start_of_packet + 4), length - 4);
```
Comment (awesomekling) [inline Kernel/Net/RTL8139NetworkAdapter.cpp]: No need for `u8*` casts here.

File: Kernel/Net/RTL8139NetworkAdapter.cpp [hunk 2]
```diff
@@ -339,31 +338,31 @@ void RTL8139NetworkAdapter::send_raw(const u8* data, size_t length)
...
     // let the card know that we've read this data
     m_rx_buffer_offset = ((m_rx_buffer_offset + length + 4 + 3) & ~3) % RX_BUFFER_SIZE;
     out16(REG_CAPR, m_rx_buffer_offset - 0x10);
     m_rx_buffer_offset %= RX_BUFFER_SIZE;
 
-    did_receive((const u8*)m_packet_buffer, length - 4);
+    did_receive((const u8*)m_packet_buffer->vaddr().as_ptr(), length - 4);
```
Comment (awesomekling) [inline Kernel/Net/RTL8139NetworkAdapter.cpp]: No need for `const u8*` cast here.

File: Kernel/Net/RTL8139NetworkAdapter.cpp [hunk 3]
```diff
@@ -339,31 +338,31 @@ void RTL8139NetworkAdapter::send_raw(const u8* data, size_t length)
 
 void RTL8139NetworkAdapter::receive()
 {
-    auto* start_of_packet = (const u8*)(low_physical_to_virtual(m_rx_buffer_addr) + m_rx_buffer_offset);
+    auto* start_of_packet = (const u8*)m_rx_buffer->vaddr().as_ptr() + m_rx_buffer_offset;
```
Comment (awesomekling) [inline Kernel/Net/RTL8139NetworkAdapter.cpp]: No need for `const u8*` cast here.

File: Kernel/Net/RTL8139NetworkAdapter.cpp [hunk 4]
```diff
@@ -320,8 +319,8 @@ void RTL8139NetworkAdapter::send_raw(const u8* data, size_t length)
         m_tx_next_buffer = (hw_buffer + 1) % 4;
     }
 
-    memcpy((void*)low_physical_to_virtual(m_tx_buffer_addr[hw_buffer]), data, length);
-    memset((void*)(low_physical_to_virtual(m_tx_buffer_addr[hw_buffer]) + length), 0, TX_BUFFER_SIZE - length);
+    memcpy((void*)m_tx_buffers[hw_buffer]->vaddr().as_ptr(), data, length);
+    memset((void*)((u8*)m_tx_buffers[hw_buffer]->vaddr().as_ptr() + length), 0, TX_BUFFER_SIZE - length);
```
Comment (awesomekling) [inline Kernel/Net/RTL8139NetworkAdapter.cpp]: No need to cast `as_ptr()`

**Comment by supercomputer7** [comment]
@awesomekling I still need to change some calls to MemoryManager. I'll let you know when I'm done with that.

**Comment by supercomputer7** [comment]
@awesomekling I think that I'm done, I had to change some lines so the network adapters will use the correct allocation helper from `MemoryManager`.

File: Kernel/VM/PhysicalRegion.cpp [hunk 5]
```diff
@@ -63,36 +64,74 @@ unsigned PhysicalRegion::finalize_capacity()
...
+    auto range = find_and_allocate_contiguous_range(m_last, m_pages, count);
+    if (!range.has_value()) {
+        // wrap back around to the start in case we missed something
+        range = find_and_allocate_contiguous_range(0, m_last, count);
+    }
+    ASSERT(range.has_value());
+    return range.value();
+}
+
+Optional<unsigned> PhysicalRegion::find_and_allocate_contiguous_range(unsigned first_page, unsigned last_page, size_t count)
```
Comment (awesomekling) [inline Kernel/VM/PhysicalRegion.cpp]: Can this function reuse what we already have in the `Bitmap` class, e.g. `Bitmap::find_longest_range_of_unset_bits`?

File: AK/Bitmap.h [hunk 6]
```diff
@@ -156,7 +156,7 @@ class Bitmap {
             if (!get(j))
                 return j;
 
-        return -1;
```
Comment (supercomputer7) [inline AK/Bitmap.h]: @awesomekling I'm still not very sure if this change is necessary.

File: AK/Bitmap.h [hunk 7]
```diff
@@ -156,7 +156,7 @@ class Bitmap {
...
+        return {};
```
Comment (supercomputer7) [inline AK/Bitmap.h]: @awesomekling I'm still not very sure if this change is necessary.

Comment (awesomekling) [reply AK/Bitmap.h]: This change is great, it's fixing a bug!
However, it's the default constructor of `Optional`, not `Bitmap`. :)

Comment (supercomputer7) [reply AK/Bitmap.h]: Oops, my bad, I'll fix it :)

**Comment by supercomputer7** [comment]
@awesomekling I tested both RTL8139 & E1000 cards. They work fine, though the RTL8139 is complaining on buffer overflow when loading big pages. That's fine for now, I assume we can fix it later.

**Comment by awesomekling** [comment]
Okay, great! Good idea to use the supervisor pages instead of writing a new physical page allocator entirely. :)

**Pull request merged**